test/mpi/dtpools/src/dtpools_internal.c

```diff
@@ -0,0 +1,1167 @@
+{
+    int i, err = DTP_SUCCESS;
+    int len, obj_idx;
+    int *type_displs = NULL;
+    int basic_type_size;
+    MPI_Aint type_offset, type_count, type_stride, type_blklen;
+    MPI_Aint size;
+    MPI_Aint extent, lb;
+    MPI_Datatype basic_type;
+    void *buf;
+    DTPI_t *dtpi = NULL;
+    char type_name[TYPE_NAME_MAXLEN] = { 0 };
+    char basic_type_name[BASIC_TYPE_NAME_MAXLEN] = { 0 };
+
+    /* allocate space for private datatype info */
+    dtpi = (DTPI_t *) malloc(sizeof(*dtpi));
+    if (!dtpi) {
+        err = DTP_ERR_OTHER;
+        fprintf(stdout, "Out of memory in %s\n", __FUNCTION__);
+        fflush(stdout);
+        goto fn_fail;
+    }
+
+    type_count = par->type_count;
+    type_displs = (int *) malloc(sizeof(*type_displs) * type_count);
+    if (!type_displs) {
+        err = DTP_ERR_OTHER;
+        fprintf(stdout, "Out of memory in %s\n", __FUNCTION__);
+        fflush(stdout);
+        goto fn_fail;
+    }
+
+    type_offset = par->type_offset;
+    type_stride = par->type_stride;
+    for (i = 0; i < type_count; i++) {
+        type_displs[i] = type_offset + type_stride * i;
+    }
+
+    /* allocate buffer */
+    basic_type = dtp->DTP_type_signature.DTP_pool_basic.DTP_basic_type;
+    err = MPI_Type_size(basic_type, &basic_type_size);
+    if (err) {
+        DTPI_Print_error(err);
+    }
+    lb = type_offset * basic_type_size;
+    extent = type_stride * basic_type_size;
+    size = lb + extent * type_count;
+    buf = (void *) malloc(size);
+    if (!buf) {
+        err = DTP_ERR_OTHER;
+        fprintf(stdout, "Out of memory in %s\n", __FUNCTION__);
+        fflush(stdout);
+        goto fn_fail;
+    }
+    memset(buf, 0, size);
+
+    /* create block indexed datatype */
+    obj_idx = par->obj_idx;
+    type_blklen = par->type_blklen;
+    err = MPI_Type_create_indexed_block(type_count, type_blklen, type_displs, basic_type,
+                                        &dtp->DTP_obj_array[obj_idx].DTP_obj_type);
+    if (err) {
+        DTPI_Print_error(err);
+    }
+    err = MPI_Type_commit(&dtp->DTP_obj_array[obj_idx].DTP_obj_type);
+    if (err) {
+        DTPI_Print_error(err);
+    }
+    dtp->DTP_obj_array[obj_idx].DTP_obj_count = 1;
+
+    /* initialize buffer with requested params */
+    par->type_offset = lb;      /* convert offset into bytes */
+    err = MPI_Type_get_name(basic_type, basic_type_name, &len);
+    if (err) {
+        DTPI_Print_error(err);
+    }
+    DTPI_Type_init_buf(par, basic_type_name, buf);
+
+    dtp->DTP_obj_array[obj_idx].DTP_obj_buf = buf;
+    dtp->DTP_obj_array[obj_idx].private_info = dtpi;
+
+    /* initialize private datatype info data */
+    dtpi->obj_type = DTPI_OBJ_TYPE__BLOCK_INDEXED;
+    dtpi->type_basic_size = basic_type_size;
+    dtpi->type_extent = extent;
+    dtpi->type_lb = lb;
+    dtpi->type_ub = lb + extent * type_count;
+    dtpi->u.block_indexed.displs = type_displs; /* freed in DTP_obj_free() */
```

Comment:
It is confusing that `extent` here is for basic type. Perhaps you can just sent `ub=lb+extent` for all datatypes if the `extent` is the output of `MPI_Type_get_extent (new_datatype)`.